From 77339a704163be09bf213c9fb6e418187be5ae77 Mon Sep 17 00:00:00 2001
From: Yu Kuai <yukuai3@huawei.com>
Date: Mon, 8 May 2023 17:04:05 +0800
Subject: [PATCH 1619/1646] [Huawei] block: protect
 blk_mq_debugfs_register/unregister_hctx() with 'debugfs_mutex'

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188711

----------------------------------------

All operations to create and remove files under 'q->debugfs_dir' should
be protected by 'q->debugfs_mutex'.

Signed-off-by: Yu Kuai <yukuai3@huawei.com>
---
 block/blk-mq-debugfs.c | 8 ++++++++
 block/blk-mq.c         | 2 ++
 2 files changed, 10 insertions(+)

diff --git a/block/blk-mq-debugfs.c b/block/blk-mq-debugfs.c
index 1e4a5ff73bd8..e2b8942311c9 100644
--- a/block/blk-mq-debugfs.c
+++ b/block/blk-mq-debugfs.c
@@ -911,6 +911,8 @@ void blk_mq_debugfs_register_hctx(struct request_queue *q,
 	char name[20];
 	int i;
 
+	lockdep_assert_held(&q->debugfs_mutex);
+
 	if (!q->debugfs_dir)
 		return;
 
@@ -925,6 +927,8 @@ void blk_mq_debugfs_register_hctx(struct request_queue *q,
 
 void blk_mq_debugfs_unregister_hctx(struct blk_mq_hw_ctx *hctx)
 {
+	lockdep_assert_held(&hctx->queue->debugfs_mutex);
+
 	if (!hctx->queue->debugfs_dir)
 		return;
 	debugfs_remove_recursive(hctx->debugfs_dir);
@@ -937,8 +941,10 @@ void blk_mq_debugfs_register_hctxs(struct request_queue *q)
 	struct blk_mq_hw_ctx *hctx;
 	int i;
 
+	mutex_lock(&q->debugfs_mutex);
 	queue_for_each_hw_ctx(q, hctx, i)
 		blk_mq_debugfs_register_hctx(q, hctx);
+	mutex_unlock(&q->debugfs_mutex);
 }
 
 void blk_mq_debugfs_unregister_hctxs(struct request_queue *q)
@@ -946,8 +952,10 @@ void blk_mq_debugfs_unregister_hctxs(struct request_queue *q)
 	struct blk_mq_hw_ctx *hctx;
 	int i;
 
+	mutex_lock(&q->debugfs_mutex);
 	queue_for_each_hw_ctx(q, hctx, i)
 		blk_mq_debugfs_unregister_hctx(hctx);
+	mutex_unlock(&q->debugfs_mutex);
 }
 
 void blk_mq_debugfs_register_sched(struct request_queue *q)
diff --git a/block/blk-mq.c b/block/blk-mq.c
index 4fdb6fcc3790..c7610a752ba8 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -2847,7 +2847,9 @@ static void blk_mq_exit_hw_queues(struct request_queue *q,
 	queue_for_each_hw_ctx(q, hctx, i) {
 		if (i == nr_queue)
 			break;
+		mutex_lock(&q->debugfs_mutex);
 		blk_mq_debugfs_unregister_hctx(hctx);
+		mutex_unlock(&q->debugfs_mutex);
 		blk_mq_exit_hctx(q, set, hctx, i);
 	}
 }
-- 
2.17.1

